content/blog/2019-11-17-chlorophyll/index.md

```diff
@@ -0,0 +1,236 @@
++++
+title = "Chlorophyll: Synthesis-Aided Compiler for Low-Power Spatial Architectures"
+extra.author = "Shaojie Xiang and Katy Voor"
+extra.bio = """
+  [Shaojie Xiang](https://github.com/Hecmay) is a 2nd year ECE PhD student researching on programming language and distributed system. 
+  [Katy Voor](https://github.com/kavoor) is a senior interested in compilers and computer architecture.
+"""
++++
+
+===
+## Introduction 
+
+Energy efficiency has been increasingly important for embedded processors and internet of things (IoT) systems. Many spatial hardware architectures have been proposed to address the power consumption problem without compromising the performance. GreenArrays 144 (GA144) is one example of the low power spatial processor. GA144 is a stack-based 18-bit processor without shared memory and fixed clock. To program GA144, users have to write an assembly-like low level programming language [[1]](#1) to define the data storage, data movement and communication logic to make use of the spatial architecture. This paper proposes a programming model and synthesis-aided compiler to compile high level C-like programs into binary executables on GA144. The compiler-optimized program can achieve comparable results with an expert written low level program.
+
+### Background 
+
+Spatial hardware is an architecture where the users are required to assign data, computations and communications explicitly to use its computing resources, storage and interconnect network. Spatial architecture is widely used in domains like High Performance Computing and image processing for its high performance and low power consumption. However, it is especially hard to program and debug due to low programmability and slow compile time. 
+
+To tackle the programmability issue, one possible solution is domain specific language (DSL) like Spatial[[2]](#2) and T2S[[3]](#3). DSL provides specialized APIs to help users realize different optimization on target hardware with much less effort. Also people use hardware templates to solve the problem. DNN/FPGA Co-Design[[4]](#4) developed a complication flow from ML workload to spatial accelerators like FPGA and CGRA by searching through template IP pools with certain constraints. Such approaches improve the programmability without sacrificing much flexibility, but still require users to have expertise for the target hardware.
```

Comment:
This part of the post has some plurality agreement errors. It might be nice to proofread again with this specifically in mind. For example, in this paragraph:

- “*a* domain specific language”, or “domain specific *languages”
- “A DSL provides”, or “DSLs provide” (not “DSL provide”)
- like FPGAs and CGRAs